File: Sort - zOthers/CountSort.java
package src.algorithms;

import java.util.Arrays;

import src.exceptions.BadArgumentTypeException;
import src.exceptions.BadRadixArgumentException;
import src.strategy.Order;

/**
 *
 * Sorting algorithm that is based on keys between a specific range. It counts  
 * the number of occurrences of each element of the array (hashing-like).
 * Afterwards, the position of each object is calculated adding the occurrences
 * at each position.
 *
 * 1) Store the number of occurrences of each object.
 * 2) Modify the count array so that each element stores the sum of the previous
 *    counts.
 * 3) Output each object from the input sequence and decrease its count by 1.
 *
 * @author: <a href="mailto:p.aceredag@gmail.com">Pablo Acereda</a>
 * @version: 1.0
 * @license: Copyright 2021 © Pablo Acereda
 * License under Apache License, Version 2.0
 *
 * @param <T> Primitive datatype or object.
 *
 */
public class CountSort<T extends Comparable<T>> extends SortingAlgorithm<T> {
	/**
	 * Sorts an array of elements using Count Sort algorithm.
     *
     * @TimeComplexity  
     * Best    -> O(n+k) <br>
     * Average -> O(n+k) <br>
     * Worst   -> O(n²) <br>
     *
     * @SpaceComplexity
     * O(1)
     *
     * @Stable
     * Yes
     *
     * @param <T> Primitive datatype or object.
     *
     * @param elements Elements to be ordered.
     * @param order Order preference.
     *
     * @return Sorted array.
     *
	 * @throws BadArgumentTypeException
	 * @throws BadRadixArgumentException 
	 */
	@Override
	@SuppressWarnings("hiding")
	public <T extends Comparable<T>> T[] sort(T[] elements, Order order) 
			throws BadRadixArgumentException,
				   BadArgumentTypeException {
		return this.sort(elements, order, false, 0);
	}

	/**
	 * Sorts an array of elements using Count Sort algorithm. Can be used as
	 * subroutine for RadixSort.
	 *
	 * @TimeComplexity
     * Best    -> O(n+k) <br>
     * Average -> O(n+k) <br>
     * Worst   -> O(n²) <br>
     *
     * @SpaceComplexity
     * O(n+k)
     *
     * @Stable
     * Yes
	 *
	 * @param <T> Primitive datatype or object.
	 *
	 * @param elements Elements to be ordered.
     * @param order Order preference.
	 * @param isRadix This algorithm can be a subroutine for RadixSort.
	 * @param exponent Only when isRadix=true. Radix/base being evaluated.
	 *
	 * @return Sorted array.
	 *
	 * @throws BadRadixArgumentException 
	 * @throws BadArgumentTypeException
	 */
	@SuppressWarnings("hiding")
	public <T extends Comparable<T>> T[] sort(T[] elements, Order order, boolean isRadix, int exponent) 
			throws BadRadixArgumentException, 
		       	   BadArgumentTypeException {

		// Make sure arguments are right
		if (isRadix && exponent == 0) {
			throw new BadRadixArgumentException("Using CountSort as part of RadixSort. Need a radix/base to proceed!");
		} else if (!isRadix && exponent != 0) {
			throw new BadRadixArgumentException("No exponent is needed, provide exponent=0 or use function without isRadix and exponent arguments!");
		}

		// No elements to order
		if (elements.length == 0 || elements == null) {
			return elements;
		}

		// Invalid inputs
		if        (!isRadix && elements[0] instanceof Float) {
			throw new BadArgumentTypeException("This implementation is not ready to sort a floating point array!");
		} else if (!isRadix && elements[0] instanceof String) {
			throw new BadArgumentTypeException("Count Sort cannot order String arrays, try with a Character array!");
		}

		int numElements = elements.length;
		int numBuckets  = (elements[0] instanceof Integer) ? 10 : 256;

		// Sorted array
		T[] output = Arrays.copyOf(elements, numElements);
		// Store the count of each type element appearance
		int[] count = new int[numBuckets];

		// Count elements appearances.
		for (T elem : elements) {
			int digit = (elem instanceof Character) ? Character.getNumericValue((char) elem) : (int) elem;
			// When using this code as subroutine for RADIX it means that each 
			// element has multiple digits.
			int bucket = isRadix ? (digit / exponent) % numBuckets : digit;
			count[bucket]++;
		}

		// Store the sum of the previous counts, symbolizing the index of the
		// output array.
		for (int i=1; i<numBuckets; i++) {
			count[i] += count[i-1];
		}

		// Build output array.
		for (int i=0; i<numElements; i++) {
			int bucket = 0;
			int j = i;

			if (isRadix) {
				j = numElements - 1 - i;
				int digit = this.numericValue(elements[j]);
				bucket = (digit / exponent) % numBuckets;
			} else {
				int digit = this.numericValue(elements[j]);
				bucket = digit;
			}
			// Move element to output array
			output[count[bucket] - 1] = elements[j];
			count[bucket]--;
		}

		// Copy array in selected order
		for (int i=0; i<numElements; i++) {
			if (order == Order.ASC) // ASCENDING
				elements[i] = output[i];
			else					// DESCENDING
				elements[i] = output[numElements - 1 - i];
		}

		return elements;
	}

	/**
	 * Transform a character to a number value.
	 *
	 * @param <T> Primitive datatype or object.
	 *
	 * @param elements Character to transform.
	 *
	 * @return Integer transformed value.
	 */
	@SuppressWarnings("hiding")
	private <T> int numericValue(T elements) {
		if (elements instanceof Character)
			return Character.getNumericValue((char) elements);
		else
			return (int) elements;
	}

}

File: Sort-zOthers/count_sort.py
from strategy import Order, BadOrderError, BadArgumentCombinationError
import sys

__author__ = "Pablo Acereda"
__copyright__ = "Copyright 2020"
__credits__ = [ "Pablo Acereda" ]

__license__ = "Apache License 2.0"
__version__ = "1.0"
__maintainer__ = "Pablo Acereda"
__email__ = "p.aceredag@gmail.com"

def sort(array:list, order:Order=Order.ASC, is_radix:bool=False, exponent:int=None) -> list:
    """Sorts a list of characters using CountingSort.

    Sorting algorithm that is based on keys between a specific range. It counts
    the number of occurences of each element of the array (hashing-like). 
    Afterwards, the position of each object is calculated adding the 
    occurrences at each position.

    1) Store the number of occurences of each object.
    2) Modify the count array so that each element stores the sum of previos
    counts.
    3) Output each object from the input sequence and decrease its count by 1.

    DISCLAIMER: 
        When using COUNTING SORT as part of RADIX SORT, take into account that 
        the output array from this call is just a partial ordering of one digit
        at a time. Therefore, those digits need to be traversed to obtain full
        ordering of the input array.
    
    DISCLAIMER 2: 
        This implementation only allows positive integers and characters 
        sorting.

    Time Complex:
        Best    -> O(n + k); where n is the number of elements in input array
        Average -> O(n + k); where k is the range of input
        Worst   -> O(n + k)
    Space Complex (Auxiliary Space): O(n + k)
    Stable: Yes

    Args:
        array (list) -- Elements to order.
        order (Order) -- Order preference (default ASCending).
        is_radix (bool) -- This algorithm can be a subroutine for RadixSort.
        exponent (int) -- Only when is_radix=True. Radix/base being evaluated.

    Returns:
        list: Ordered elements.
    """
    # Exists the program if the ordering is not valid. 
    if (order not in [Order.ASC, Order.DESC]):
        raise BadOrderError("Not Valid Ordering Preference")

    # When using radix, exponent parameter is necessary
    if is_radix and not exponent:
        raise BadArgumentCombinationError("When using this argument as part of RadixSort, an exponent is also needed!")
    elif not is_radix and exponent:
        raise BadArgumentCombinationError("Exponent argument is not needed when not using RadixSort!")

    # Empty list
    if not array: 
        return []

    # This algorithm cannot order Strings
    if (isinstance(array[0], float)): 
        raise ValueError("This sorting algorithm does not take floating point numbers!")

    num_elems = len(array)
    # Dependending on type of list the number of buckets may vary
    # STR -> 256
    # INT -> 10
    is_str = isinstance(array[0], str)
    num_buckets = 256 if is_str else 10

    # Sorted array
    output = [0 for i in range(num_elems)]

    # Storage for the count of individual characters 
    count  = [0 for i in range(num_buckets)]

    # Count the number of appearences of each character
    for d in array:
        # Characters need to be converted to integers to be counted
        digit = ord(d) if is_str else d
        # Index to where add the count
        # When using RADIX, it is given that each value has multiple digits
        bucket = int((digit / exponent) % num_buckets) if is_radix else digit
        count[bucket] += 1

    # Each index stores the sum of the previous counts, symbolizing the index
    # of the output array
    for i in range(1, num_buckets):
        count[i] += count[i-1]

    # Build the output array
    for i in range(num_elems):
        bucket = None
        j = i
        # COUNT sort might be used as subroutine for RADIX sort
        if is_radix:
            j = (num_elems - 1) - i
            # Characters need to be converted to integers to be counted
            digit = ord(array[j]) if is_str else array[j]
            # Where to place the current element
            bucket = int((digit / exponent) % num_buckets)
        else:
            # Characters need to be converted to integers to be counted
            digit = ord(array[j]) if is_str else array[j]
            # Where to place the current element
            bucket = digit
        # Move element to output array
        output[count[bucket] - 1] = array[j] 
        count[bucket] -= 1

    # Copy sorted elements back to array, this also allows to sort the array on
    # descending order
    for i in range(num_elems):
        if (order == Order.ASC): # ASCENDING
            array[i] = output[i]
        else:                    # DESCENDING
            array[i] = output[(num_elems - 1) - i]

    return array
